akka-actor-typed/src/main/scala/akka/actor/typed/javadsl/StashBuffer.scala

```diff
@@ -0,0 +1,218 @@
+
+  /**
+   * Check if the message buffer is empty.
+   *
+   * @return if the buffer is empty
+   */
+  def isEmpty: Boolean
+
+  /**
+   * Check if the message buffer is not empty.
+   *
+   * @return if the buffer is not empty
+   */
+  def nonEmpty: Boolean
+
+  /**
+   * How many elements are in the message buffer.
+   *
+   * @return the number of elements in the message buffer
+   */
+  def size: Int
+
+  /**
+   * @return `true` if no more messages can be added, i.e. size equals the capacity of the stash buffer
+   */
+  def isFull: Boolean
+
+  /**
+   * Add one element to the end of the message buffer. Note that this class is
+   * immutable so the returned instance contains the added message.
+   *
+   * @param message the message to buffer, must not be `null`
+   * @return this message buffer
+   * @throws  `StashOverflowException` is thrown if the buffer [[MutableStashBuffer#isFull]].
+   */
+  def stash(message: T): ImmutableStashBuffer[T]
+
+  /**
+   * Remove the first element of the message buffer. Note that this class is
+   * immutable so the head element is removed in the returned instance.
+   *
+   * @throws `NoSuchElementException` if the buffer is empty
+   */
+  def dropHead(): ImmutableStashBuffer[T]
+
+  /**
+   * Remove the first `numberOfMessages` of the message buffer. Note that this class is
+   * immutable so the elements are removed in the returned instance.
+   */
+  def drop(numberOfMessages: Int): ImmutableStashBuffer[T]
+
+  /**
+   * Return the first element of the message buffer.
+   *
+   * @return the first element or throws `NoSuchElementException` if the buffer is empty
+   * @throws `NoSuchElementException` if the buffer is empty
+   */
+  def head: T
+
+  /**
+   * Iterate over all elements of the buffer and apply a function to each element.
+   *
+   * @param f the function to apply to each element
+   */
+  def forEach(f: Consumer[T]): Unit
+
+  /**
+   * Process all stashed messages with the `behavior` and the returned
+   * [[Behavior]] from each processed message.
+   */
+  def unstashAll(ctx: ActorContext[T], behavior: Behavior[T]): Behavior[T]
+
+  /**
+   * Process `numberOfMessages` of the stashed messages with the `behavior`
+   * and the returned [[Behavior]] from each processed message.
+   */
+  def unstash(ctx: ActorContext[T], behavior: Behavior[T], numberOfMessages: Int, wrap: JFunction[T, T]): Behavior[T]
+
+}
+
+object MutableStashBuffer {
+
+  /**
+   * Create an empty message buffer.
+   *
+   * @param capacity the buffer can hold at most this number of messages
+   * @return an empty message buffer
+   */
+  def create[T](capacity: Int): MutableStashBuffer[T] =
+    MutableStashBufferImpl[T](capacity)
+}
+
+/**
+ * A non thread safe mutable message buffer that can be used to buffer messages inside actors.
+ *
+ * The buffer can hold at most the given `capacity` number of messages.
+ */
+@DoNotInherit abstract class MutableStashBuffer[T] {
```

Comment:
Alternatively users could just use the Immutable one with a var. But i guess this is more error prone for users who chose the mutable behavior